File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = maxlen= 0
        charset = set()
        for right in range(len(s)):
            while s[right] in charset :
                charset.remove(s[left])
                left +=1
            charset.add(s[right])
            maxlen = max(maxlen, right - left +1)
        return maxlen

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left=0;
        int maxlen = 0;
        HashSet<Character> charset = new HashSet<>();
        for(int right=0;right<s.length();right++){
            while(charset.contains(s.charAt(right))){
                charset.remove(s.charAt(left));
                left++;
            }
            charset.add(s.charAt(right));
            maxlen = Math.max(maxlen, right - left +1);
        }
        return maxlen;
    }
}